### PR TITLE
Metal Support

### PR DESCRIPTION
**Pull request by grolu**
**What this PR does / why we need it**:
This adds support for `metal` additional infrastructure kind.

Screen Shot 2020-03-04 at 13 26 44
Screen Shot 2020-03-04 at 13 27 10

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```improvement user
Added support for Metal Pod 
```

### CODE REVIEW

File: frontend/public/static/background_metal.svg [hunk 1]
```diff
@@ -0,0 +1,62 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<svg width="720px" height="200px" viewBox="0 0 720 200" version="1.1" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink">
+    <!-- Generator: Sketch 61.2 (89653) - https://sketch.com -->
+    <title>vsphere_background</title>
+    <desc>Created with Sketch.</desc>
```
Comment (petersutter) [inline frontend/public/static/background_metal.svg]:
```suggestion
```

File: frontend/src/assets/metal.svg [hunk 2]
```diff
@@ -0,0 +1,26 @@
+<?xml version="1.0" encoding="utf-8"?>
+<!-- Generator: Adobe Illustrator 24.1.1, SVG Export Plug-In . SVG Version: 6.00 Build 0)  -->
```
Comment (petersutter) [inline frontend/src/assets/metal.svg]:
```suggestion
```

File: frontend/src/utils/index.js [hunk 3]
```diff
@@ -599,6 +586,20 @@ export function generateWorker (availableZones, cloudProfileName, region) {
   return worker
 }
 
+export function isZonedCluster ({ cloudProviderKind, shootSpec }) {
+  switch (cloudProviderKind) {
+    case 'azure':
+      if (!shootSpec) {
```
Comment (petersutter) [inline frontend/src/utils/index.js]: make more explicit

File: frontend/src/components/NewShoot/NewShootInfrastructureDetails.vue [hunk 4]
```diff
@@ -190,6 +270,31 @@ const validations = {
       return this.infrastructureKind === 'vsphere'
     }),
     includesKey: includesIfAvailable('default', 'allLoadBalancerClassNames')
+  },
+  partitionID: {
+    required: requiredIf(function () {
+      return this.infrastructureKind === 'metal'
```
Comment (petersutter) [inline frontend/src/components/NewShoot/NewShootInfrastructureDetails.vue]: We need to check if we can move provider specific code to an own component without having to duplicate things. This way we can introduce additional infra providers without the fear of breaking existing providers. WDYT?
Of course, nothing we would do within this PR.

File: frontend/src/components/NewShoot/NewShootInfrastructureDetails.vue [hunk 5]
```diff
@@ -324,6 +442,18 @@ export default {
...
       return this.loadBalancerClassNamesByCloudProfileName(this.cloudProfileName)
     },
+    partitionIDs () {
+      return this.partitionIDsByCloudProfileNameAndRegion({ cloudProfileName: this.cloudProfileName, region: this.region })
+    },
+    firewallImages () {
+      return this.firewallImagesByCloudProfileName(this.cloudProfileName)
+    },
+    firewallSizes () {
+      return map(this.firewallSizesByCloudProfileNameAndRegionAndZones({ cloudProfileName: this.cloudProfileName, region: this.region, zones: [ this.partitionID ] }), 'name')
```
Comment (petersutter) [inline frontend/src/components/NewShoot/NewShootInfrastructureDetails.vue]:
```suggestion
      const cloudProfileName = this.cloudProfileName
      const region = this.region
      const zones = [this.partitionID]
      const firewallSizes = this.firewallSizesByCloudProfileNameAndRegionAndZones({ cloudProfileName, region, zones })
      return map(firewallSizes, 'name')
```
suggestion not checked for linting / correctness

File: frontend/src/components/NewShoot/NewShootInfrastructureDetails.vue (hunk 4 above)
Comment (grolu) [reply frontend/src/components/NewShoot/NewShootInfrastructureDetails.vue]: Yes, we already discussed this briefly. I think this would be feasible.

File: frontend/src/utils/createShoot.js [hunk 6]
```diff
@@ -23,9 +23,52 @@ import find from 'lodash/find'
...
+              enabled: true
+            }
+          }
+        },
+        kubernetes: {
+          kubeControllerManager: {
+            nodeCIDRMaskSize: 23
+          },
+          kubelet: {
+            maxPods: 510
```
Comment (petersutter) [inline frontend/src/utils/createShoot.js]: is there a plan to remove this in the future (from the dashboard)?

File: frontend/src/components/ShootWorkers/ManageWorkers.vue [hunk 7]
```diff
@@ -184,20 +185,24 @@ export default {
...
+      this.isNewCluster = isNewCluster
     }
   },
   mounted () {
     if (this.userInterActionBus) {
       this.userInterActionBus.on('updateCloudProfileName', cloudProfileName => {
         this.internalWorkers = []
         this.cloudProfileName = cloudProfileName
+        const cloudProfile = this.cloudProfileByName(cloudProfileName)
+        // shootspec not required, userInterActionBus callback only called for new clusters
```
Comment (petersutter) [inline frontend/src/components/ShootWorkers/ManageWorkers.vue]: improve comment

File: frontend/src/components/ShootWorkers/WorkerConfiguration.vue [hunk 8]
```diff
@@ -94,7 +85,7 @@ export default {
       const workers = cloneDeep(this.shootWorkerGroups)
       const zonesNetworkConfiguration = get(this.shootItem, 'spec.provider.infrastructureConfig.networks.zones')
 
-      this.$refs.manageWorkers.setWorkersData({ workers, cloudProfileName: this.shootCloudProfileName, region: this.shootRegion, zonesNetworkConfiguration, zonedCluster: this.isZonedCluster })
+      this.$refs.manageWorkers.setWorkersData({ workers, cloudProfileName: this.shootCloudProfileName, region: this.shootRegion, zonesNetworkConfiguration, zonedCluster: isZonedCluster({ cloudProviderKind: this.shootCloudProviderKind, shootSpec: this.shootSpec })})
```
Comment (petersutter) [inline frontend/src/components/ShootWorkers/WorkerConfiguration.vue]:
```suggestion
      const zonedCluster = isZonedCluster({ cloudProviderKind: this.shootCloudProviderKind, shootSpec: this.shootSpec })
      this.$refs.manageWorkers.setWorkersData({ workers, cloudProfileName: this.shootCloudProfileName, region: this.shootRegion, zonesNetworkConfiguration, zonedCluster })
```

Comment (petersutter) [reply frontend/src/components/ShootWorkers/WorkerConfiguration.vue]: extract to variable

File: frontend/src/pages/NewShoot.vue [hunk 9]
```diff
@@ -224,16 +225,21 @@ export default {
...
-      set(shootResource, 'spec.secretBindingName', get(secret, 'metadata.bindingName'))
       const oldInfrastructureKind = get(shootResource, 'spec.provider.type')
       if (oldInfrastructureKind !== infrastructureKind) {
         // Infrastructure changed
-        set(shootResource, 'spec.provider', getProviderTemplate(infrastructureKind))
+        set(shootResource, 'spec', getSpecTemplate(infrastructureKind))
       }
+      set(shootResource, 'spec.cloudProfileName', cloudProfileName)
+      set(shootResource, 'spec.region', region)
+      set(shootResource, 'spec.secretBindingName', get(secret, 'metadata.bindingName'))
```
Comment (petersutter) [inline frontend/src/pages/NewShoot.vue]: why were these lines added?

File: frontend/src/pages/NewShoot.vue [hunk 10]
```diff
@@ -345,7 +378,7 @@ export default {
       this.$refs.clusterDetails.setDetailsData({ name, kubernetesVersion, purpose, secret, cloudProfileName, updateK8sMaintenance: k8sUpdates })
 
       const workers = get(shootResource, 'spec.provider.workers')
-      this.$refs.manageWorkers.setWorkersData({ workers, cloudProfileName, region, updateOSMaintenance: osUpdates })
+      this.$refs.manageWorkers.setWorkersData({ workers, cloudProfileName, region, updateOSMaintenance: osUpdates, zonedCluster: isZonedCluster({ cloudProviderKind: infrastructureKind, isNewCluster: true }), isNewCluster: true })
```
Comment (petersutter) [inline frontend/src/pages/NewShoot.vue]: extract variables to make it better readable

File: frontend/src/store/index.js [hunk 11]
```diff
@@ -341,6 +342,48 @@ const getters = {
       return get(cloudProfile, 'data.providerConfig.constraints.loadBalancerConfig.classes')
     }
   },
+  partitionIDsByCloudProfileNameAndRegion (state, getters) {
+    return ({ cloudProfileName, region }) => {
+      // Partion IDs equal zones for metal infrastructure
+      const cloudProfile = getters.cloudProfileByName(cloudProfileName)
+      if (cloudProfile.metadata.cloudProviderKind !== 'metal') {
```
Comment (petersutter) [inline frontend/src/store/index.js]: can we be sure that cloudProfile / cloudProfile.metadata. is always there? same applies for methods introduced below

File: frontend/src/components/ShootWorkers/WorkerConfiguration.vue (hunk 8 above)
Comment (grolu) [reply frontend/src/components/ShootWorkers/WorkerConfiguration.vue]: done

File: frontend/src/store/index.js (hunk 11 above)
Comment (grolu) [reply frontend/src/store/index.js]: I would say yes if we don't have an incorrect state?

File: frontend/src/components/NewShoot/NewShootInfrastructureDetails.vue (hunk 4 above)
Comment (holgerkoser) [reply frontend/src/components/NewShoot/NewShootInfrastructureDetails.vue]: 👍

File: frontend/src/dialogs/SecretDialogAzure.vue [hunk 12]
```diff
@@ -21,7 +21,7 @@ limitations under the License.
     :dataValid="valid"
     :secret="secret"
     cloudProviderKind="azure"
-    color="blue"
+    color="blue darken-1"
```
Comment (holgerkoser) [inline frontend/src/dialogs/SecretDialogAzure.vue]: Nothing to change here but we should have constants for infrastructureProvider colors and textColors.

File: frontend/src/dialogs/SecretDialogGcp.vue [hunk 13]
```diff
@@ -36,7 +36,7 @@ limitations under the License.
             color="green"
             box
             v-model="serviceAccountKey"
-            :label="serviceAccountKeyLabel"
+            label="Service Account Key"
```
Comment (holgerkoser) [inline frontend/src/dialogs/SecretDialogGcp.vue]: Do we keep the serviceAccountKey?

File: frontend/src/dialogs/SecretDialogMetal.vue [hunk 14]
```diff
@@ -0,0 +1,158 @@
...
+      type: Boolean,
+      required: true
+    },
+    secret: {
+      type: Object
+    }
+  },
+  data () {
+    return {
+      apiHMAC: undefined,
```
Comment (holgerkoser) [inline frontend/src/dialogs/SecretDialogMetal.vue]: Use one kind of case

File: frontend/src/store/index.js (hunk 11 above)
Comment (grolu) [reply frontend/src/store/index.js]: implemented safe access

File: frontend/src/dialogs/SecretDialogMetal.vue (hunk 14 above)
Comment (grolu) [reply frontend/src/dialogs/SecretDialogMetal.vue]: renamed as requested
